pymc_bart/tree.py

```diff
@@ -52,164 +106,161 @@ class Tree:
         self.tree_structure[index] = node
-        if node.is_leaf_node():
-            self.idx_leaf_nodes.append(index)
+        if node.is_leaf_node() and self.idx_leaf_nodes is not None:
+            # self.idx_leaf_nodes.append(index)
+            self.idx_leaf_nodes = np.append(self.idx_leaf_nodes, index)
 
-    def grow_leaf_node(self, current_node, selected_predictor, split_value, index_leaf_node):
+    def grow_leaf_node(
+        self, current_node: Node, selected_predictor: int, split_value: float, index_leaf_node: int
+    ) -> None:
         current_node.value = split_value
         current_node.idx_split_variable = selected_predictor
         current_node.idx_data_points = None
-        self.idx_leaf_nodes.remove(index_leaf_node)
+        if self.idx_leaf_nodes is not None:
+            # self.idx_leaf_nodes.remove(index_leaf_node)
+            self.idx_leaf_nodes = np.setdiff1d(self.idx_leaf_nodes, index_leaf_node)
 
-    def trim(self):
-        tree = {
+    def trim(self) -> "Tree":
+        tree: Dict[int, Node] = {
             k: Node(v.index, v.value, None, v.idx_split_variable)
             for k, v in self.tree_structure.items()
         }
-        return Tree(tree, None, None)
+        return Tree(tree_structure=tree, idx_leaf_nodes=None, output=np.array([-1]))
 
-    def get_split_variables(self):
+    def get_split_variables(self) -> Generator[int, None, None]:
         for node in self.tree_structure.values():
             if node.is_split_node():
                 yield node.idx_split_variable
 
-    def _predict(self):
+    def _predict(self) -> npt.NDArray[np.float_]:
         output = self.output
-        for node_index in self.idx_leaf_nodes:
-            leaf_node = self.get_node(node_index)
-            output[leaf_node.idx_data_points] = leaf_node.value
+
+        if self.idx_leaf_nodes is not None:
+            for node_index in self.idx_leaf_nodes:
+                leaf_node = self.get_node(node_index)
+                output[leaf_node.idx_data_points] = leaf_node.value
         return output.T
 
-    def predict(self, x, excluded=None):
+    def predict(
+        self, x: npt.NDArray[np.float_], excluded: Optional[npt.NDArray[np.int_]] = None
+    ) -> npt.NDArray[np.float_]:
         """
         Predict output of tree for an (un)observed point x.
 
         Parameters
         ----------
-        x : numpy array
+        x : npt.NDArray[np.float_]
             Unobserved point
-        excluded: list
-                Indexes of the variables to exclude when computing predictions
+        excluded: Optional[npt.NDArray[np.int_]]
+            Indexes of the variables to exclude when computing predictions
 
         Returns
         -------
-        float
+        npt.NDArray[np.float_]
             Value of the leaf value where the unobserved point lies.
         """
         if excluded is None:
-            excluded = []
+            excluded = np.array([])
```

Comment:
can we keep the list? I don't think we need an array.

Reply:
Not for operations like appending. But yes for typical "numerical" operations like computing trigonometric operations or vector operations. 

Sometimes also makes sense to write code in "pure python" and then use numba to speed-up the code. In many instances that can be 2x or faster than numpy/scipy (we have a few examples of that in PyMC-BART). Also when your function is very specialized like if all you are going to do is to perform operations with points in 3D space, an ad-hoc "dot" function can be faster than "np.dot".

Reply:
Fixed in https://github.com/pymc-devs/pymc-bart/pull/78/commits/407e4a6d2619c7d71d4c250ab600db61b2b9e071